File: 2018/3/main.py
from collections import Counter

squares = {}
claims = {}

with open('input.txt') as f:
    for line in f.readlines():
        _id, _claim = line.split("@")
        offset, area = _claim.split(":")
        offset = [int(x) for x in offset.split(",")]
        area = [int(x) for x in area.split("x")]
        
        claim = set()
        for x in range(offset[0], offset[0] + area[0]):
            for y in range(offset[1], offset[1] + area[1]):
                square = (x, y)
                claim.add(square)
                if square not in squares:
                    squares[square] = 1
                else:
                    squares[square] += 1
        claims[_id] = claim

answer_1 = 0
for square, count in squares.items():
    if count > 1:
        answer_1 += 1

print(answer_1)

for _id, s in claims.items():
    for square in s:
        if squares.get(square) != 1:
            break
    else:
        print(_id)
        break

File: 2018/3/main.hs
import qualified Data.Set     as Set
import Data.List
import Data.List.Split (splitOn)
import Data.Maybe      (fromJust)

plotCoords :: (Int, Int) -> (Int, Int) -> [(Int, Int)]
plotCoords (offsetX, offsetY) (areaX, areaY) = [(x, y) | x <-rangeX, y <-rangeY]
    where
        rangeX = [offsetX..(offsetX+areaX-1)]
        rangeY = [offsetY..(offsetY+areaY-1)]

toIntPair :: [String] -> (Int, Int)
toIntPair [a, b] = (read a ::Int, read b ::Int)

plotClaim :: String -> [(Int, Int)]
plotClaim x = plotCoords offset area
    where
        offset = toIntPair (splitOn "," (head split))
        area   = toIntPair (splitOn "x" (last split))
        split  = splitOn ":" x

convertClaim :: String -> (String, [(Int, Int)])
convertClaim x = (head y, plotClaim $ last y)
    where
        y = splitOn "@" x

dupeSet :: [(Int, Int)] -> Set.Set (Int, Int)
dupeSet coords = Set.fromList $ concat $ filter ((>1) . length) (group $ sort $ coords)

checkFree :: Set.Set (Int, Int) -> (String, [(Int, Int)]) -> Bool
checkFree dupes (id, coords) = if Set.disjoint (Set.fromList coords) dupes then True else False

main = do
    file_lines <- fmap lines (readFile "input.txt")
    let claims = map convertClaim file_lines
    let dupes = dupeSet (concat (map snd claims))
    putStrLn $ show $ length $ dupes
    putStrLn $ show $ fst $ fromJust $ find (checkFree dupes) claims
